### PR TITLE
Istio Ingress Informers 

### PR DESCRIPTION
**Pull request by rheaj99**
This PR adds informers for Istio CRD and adds a IstioClient to the main function

### CODE REVIEW

File: pkg/annotations/ingress_annotations.go [hunk 1]
```diff
@@ -43,6 +43,10 @@ const (
 	// ApplicationGatewayIngressClass defines the value of the `IngressClassKey` annotation that will tell the ingress controller
 	// whether it should act on this ingress resource or not.
 	ApplicationGatewayIngressClass = "azure/application-gateway"
+
+	// ApplicationGatewayIstioIngressClass defines the value of the `IngressClassKey` annotation that will tell the ingress controller whether
+	// this ingress resource is from Istio or not.
+	ApplicationGatewayIstioIngressClass = "istio-ingress"
```
Comment (draychev) [inline pkg/annotations/ingress_annotations.go]: Here is what a simple definition of an Istio Gateway might look like:
```yaml
kubectl apply -f - <<EOF
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: httpbin-gateway
spec:
  selector:
    istio: ingressgateway # use Istio default gateway implementation
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
    hosts:
    - "httpbin.example.com"
```

With the proposed annotation here, we'd see the following added to the YAML above:
```yaml
metadata:
    annotations:
        kubernetes.io/ingress.class: istio-ingress
```

I don't think that's quite right.

I think that we can re-use pieces of our existing `ApplicationGatewayIngressClass` annotation but put it in the Istio namespace (to be semantically correct... not that that string really matters).

How about:
```yaml
kubectl apply -f - <<EOF
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: httpbin-gateway
  annotations:
    # option A - create a new tag - namespaced within istio.io
    istio.io/ingress.class: azure/application-gateway
    # option B - reuse the existing tag we have
    kubernetes.io/ingress.class: azure/application-gateway
spec:
  selector:
    istio: ingressgateway # use Istio default gateway implementation
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
    hosts:
    - "httpbin.example.com"
```

File: cmd/appgw-ingress/main.go [hunk 2]
```diff
@@ -96,11 +97,12 @@ func main() {
 
 	apiConfig := getKubeClientConfig()
 	kubeClient := kubernetes.NewForConfigOrDie(apiConfig)
+	istioCrdClient := istio.NewForConfigOrDie(apiConfig)
```
Comment (akshaysngupta) [inline cmd/appgw-ingress/main.go]: can order the clients instantiation in order of kubeClient, crdClient, istioCrdClient

File: cmd/appgw-ingress/main.go [hunk 3]
```diff
@@ -96,11 +97,12 @@ func main() {
...
 	namespaces := getNamespacesToWatch(env.WatchNamespace)
 	validateNamespaces(namespaces, kubeClient) // side-effect: will panic on non-existent namespace
 	glog.Info("Ingress Controller will observe the following namespaces:", strings.Join(namespaces, ","))
 	crdClient := versioned.NewForConfigOrDie(apiConfig)
-	k8sContext := k8scontext.NewContext(kubeClient, crdClient, namespaces, *resyncPeriod)
+	k8sContext := k8scontext.NewContext(istioCrdClient, kubeClient, crdClient, namespaces, *resyncPeriod)
```
Comment (akshaysngupta) [inline cmd/appgw-ingress/main.go]: same ordering here.

File: pkg/k8scontext/context.go [hunk 4]
```diff
@@ -40,13 +42,15 @@ func NewContext(kubeClient kubernetes.Interface, crdClient versioned.Interface,
...
-		Ingress:   informerFactory.Extensions().V1beta1().Ingresses().Informer(),
-		Pods:      informerFactory.Core().V1().Pods().Informer(),
-		Secret:    informerFactory.Core().V1().Secrets().Informer(),
-		Service:   informerFactory.Core().V1().Services().Informer(),
+		Endpoints:    informerFactory.Core().V1().Endpoints().Informer(),
+		Ingress:      informerFactory.Extensions().V1beta1().Ingresses().Informer(),
+		Pods:         informerFactory.Core().V1().Pods().Informer(),
+		Secret:       informerFactory.Core().V1().Secrets().Informer(),
+		Service:      informerFactory.Core().V1().Services().Informer(),
+		IstioGateway: istioGwy.Networking().V1alpha3().Gateways().Informer(),
```
Comment (akshaysngupta) [inline pkg/k8scontext/context.go]: can you move IstioGateway after AzureIngressProhibitedLocation leaving a line ?

File: pkg/k8scontext/context.go [hunk 5]
```diff
@@ -40,13 +42,15 @@ func NewContext(kubeClient kubernetes.Interface, crdClient versioned.Interface,
 	}
 	informerFactory := informers.NewSharedInformerFactoryWithOptions(kubeClient, resyncPeriod, options...)
 	crdInformerFactory := externalversions.NewSharedInformerFactoryWithOptions(crdClient, resyncPeriod, crdOptions...)
+	istioGwy := istio_externalversions.NewSharedInformerFactoryWithOptions(istioCrdClient, resyncPeriod)
```
Comment (akshaysngupta) [inline pkg/k8scontext/context.go]: istioCrdInformerFactory

File: pkg/k8scontext/types.go [hunk 6]
```diff
@@ -15,6 +15,7 @@ type InformerCollection struct {
 	Secret                         cache.SharedIndexInformer
 	Service                        cache.SharedIndexInformer
 	Namespace                      cache.SharedIndexInformer
+	IstioGateway                   cache.SharedIndexInformer
```
Comment (akshaysngupta) [inline pkg/k8scontext/types.go]: same here

**Pull request merged**